### ISSUE TITLE
[Enhancement]: Expand API parameters when fetching products to allow for more granularity

### ISSUE THREAD

**Issue by iamgabrielma** [woothemes/woocommerce#50462]
### Describe the solution you'd like

The current version of the `products` [endpoint](https://woocommerce.github.io/woocommerce-rest-api-docs/#list-all-products) is somewhat limited to fetching either all products, fetching all known IDs, or fetch by very specific product properties. It would be great to expand the API to allow more options, like for example we do for `status` or `type`:
```

- status: Limit result set to products assigned a specific status. Options: any, draft, pending, private and publish. Default is any.
- type: Limit result set to products assigned a specific type. Options: simple, grouped, external and variable.
```

When other product properties need to be used this could bring some issues with pagination due to the batch limit, for example, if we want to fetch all `virtual` products, at the moment we would need to fetch all products, and then filter them further in the client side by which ones are virtual and which ones aren't.

This applies to more properties, eg: `downloadable`, `purchasable`, etc, ...  or to some properties that might accept multiple values but the API only allows one, for example in order to fetch only `published AND private` products we would need to perform 2 requests, since the API doesn't accept an array of values:

```
status: String
```

### Additional context

For additional context, this came up during the Woo POS POC project here: pdfdoF-5d5-p2#comment-6290

**Comment by AnirudhBhat** [comment]
The ability to filter products via the API would be highly beneficial for supporting non-simple product types pdfdoF-5yx-p2 project. We have a requirement to filter products by multiple types, such as Simple and Variable products, to meet specific use cases.

**Comment by AnirudhBhat** [comment]
Here's a detailed P2 post for more context pdfdoF-5Ej-p2

**Comment by malinajirka** [comment]
Hi @barryhughes, this change will become a blocker for the POS project as it breaks pagination (more info on the references P2 post - pdfdoF-5d5-p2#comment-6290). Any chance we could increase a priority of it?

**Comment by barryhughes** [comment]
I think we can.

(Edit: just for clarity, though, I'm not modifying the priority label as `priority: high` has its own special meaning.)

**Comment by AnirudhBhat** [comment]
:wave: Hi @barryhughes, I wanted to check on the timeline for the fix. My goal is to wrap up the client-side development for the non-simple product types project by the end of this week. If testing proceeds smoothly, we’re targeting November 22nd to complete the MVP scope.

**Comment by ctarda** [comment]
@AnirudhBhat We are aiming to start working on this issue this week. I don't think we will be done by November 22nd though

**Comment by ctarda** [comment]
And one more thing @AnirudhBhat 😊 @albarin will DRI this project on the API side

**Comment by albarin** [comment]
@AnirudhBhat @iamgabrielma would you help me clarify if I'm understanding correctly what's needed? 🙏 
According to this pdfdoF-5Ej-p2 you need the `include_types` and `exclude_types` parameters in the `/products` endpoint.
You also mentioned the `status` parameter on the issue description, do you also need it to accept multiple statutes instead of just one, right?

**Comment by AnirudhBhat** [comment]
>We are aiming to start working on this issue this week. I don't think we will be done by November 22nd though

@ctarda Thanks a ton for prioritising this change. I really appreciate it 🙏 May I know the guesstimate you are looking at for this change? Thanks!

**Comment by AnirudhBhat** [comment]
>would you help me clarify if I'm understanding correctly what's needed? 🙏
According to this pdfdoF-5Ej-p2 you need the include_types and exclude_types parameters in the /products endpoint.
You also mentioned the status parameter on the issue description, do you also need it to accept multiple statutes instead of just one, right?

That's correct. As mentioned in the P2 post, it would be better if we add a new field for `status` as well instead of modifying the existing `status` field as it breaks existing implementations.

One more thing I wanted to bring to your attention is that we also want to filter product by virtual, downloadable, or purchasable which which are not actual types but just flags on Simple products.

To Summarise:
1. `include_types` and `exclude_types`: Enable filtering by specific product types, such as Simple, Variable, and Grouped.
2. Status Filtering: Introduce a separate field for status filtering, allowing filtering by status values without modifying the existing status field to preserve compatibility.
3. Flag-Based Filtering: Add filtering capabilities for flags on Simple products, such as virtual, downloadable, or purchasable, using the same field as in point 1. These flags aren’t distinct types but are important attributes for filtering.

@malinajirka  @iamgabrielma Please add if I'm missing anything here. Thanks!
